File: Src/Engine/Editor/Resource/Items/ContentItem.cpp

#include "ContentItem.h"

#include "ContentFolder.h"
#include "Core/Platform/FileSystem.h"
#include "Editor/EditorIcons.h"
#include "Editor/GUI/Drag/DragItems.hpp"
#include "Editor/Modules/UIModule.h"
#include "Editor/Resource/ContentView.h"
#include "Editor/Resource/Thumbnails/Thumbnails.h"
#include "Runtime/Render/2D/Render2D.h"
#include "Runtime/UI/GUI/Style.h"

namespace SE::Editor
{
	Rectangle ContentItem::__GetTextRectangle()
	{
		// Skip when hidden
		if (!Visible)
			return Rectangle::Empty;

		ContentView* view = TypeTryCast<ContentView>(Parent);
		Float2 size = Size;
		ContentViewType viewType = ContentViewType::Tiles;
		if (view != nullptr)
		{
			viewType = view->ViewType;
		}
		switch (viewType)
		{
			case ContentViewType::Tiles:
			{
				float textHeight = DefaultTextHeight * size.x / DefaultWidth;
				return Rectangle(0, size.y - textHeight, size.x, textHeight);
			}
			case ContentViewType::List:
			{
				float thumbnailSize = size.y - 2 * DefaultMarginSize;
				float textHeight = Math::Min(size.y, 24.0f);
				return Rectangle(thumbnailSize + DefaultMarginSize * 2, (size.y - textHeight) * 0.5f, size.x - textHeight - DefaultMarginSize * 3.0f, textHeight);
			}
		}

		return Rectangle::Empty;
	}

	ContentFolder* ContentItem::__GetParentFolder()
	{
		return _parentFolder;
	};

	void ContentItem::__SetParentFolder(ContentFolder* value)
	{
		if (_parentFolder == value)
			return;

		// Remove from old
		if (_parentFolder != nullptr)
		{
			_parentFolder->Children.Remove(this);
		}

		// Link
		_parentFolder = value;

		// Add to new
		if (_parentFolder != nullptr)
		{
			_parentFolder->Children.Add(this);
		}

		OnParentFolderChanged();
	};

	void ContentItem::UpdatePath(String value)
	{
		// Set path
		FileSystem::NormalizePath(value);
		Path = value;
		FileName = FileSystem::GetFileName(value);
		ShortName = FileSystem::GetFileNameWithoutExtension(value);

		// Fire event
		OnPathChanged();
		for (int i = 0; i < _references.Count(); i++)
		{
			_references[i]->OnItemRenamed(this);
		}
	}

	void ContentItem::RefreshThumbnail()
	{
		// Skip if item has default thumbnail
		if (HasDefaultThumbnail)
			return;

		Thumbnails* thumbnails = EditorApp::Ins().uiModule->thumbnails;

		// Delete old thumbnail and remove it from the cache
		thumbnails->DeletePreview(this);

		// Request new one (if need to)
		if (_references.Count() > 0)
		{
			thumbnails->RequestPreview(this);
		}
	}

	void ContentItem::UpdateTooltipText()
	{
		/*StringBuilder sb = StringBuilder();
		OnBuildTooltipText(sb);
		if (sb.Length() != 0 && sb[sb.Length() - 1] == '\n')
		{
			// Remove new-line from end
			int sub = 1;
			if (sb.Length() != 1 && sb[sb.Length() - 2] == '\r')
				sub = 2;
			sb.Length() -= sub;
		}
		TooltipText = sb.ToString();*/
	}

	ContentItem* ContentItem::Find(String path)
	{
		return Path == path ? this : nullptr;
	}

	bool ContentItem::Find(ContentItem* item)
	{
		return this == item;
	}

	ContentItem* ContentItem::Find(UID id)
	{
		return nullptr;
	}

	void ContentItem::DrawThumbnail(Rectangle rectangle)
	{
		// Draw shadow
		/*if (DrawShadow())
		{
			const float thumbnailInShadowSize = 50.0f;
			var shadowRect = rectangle.MakeExpanded((DefaultThumbnailSize - thumbnailInShadowSize) * rectangle.Width / DefaultThumbnailSize * 1.3f);
			if (!_shadowIcon.IsValid)
				_shadowIcon = Editor.Instance.Icons.AssetShadow128;
			Render2D::DrawSprite(_shadowIcon, shadowRect);
		}*/

		// Draw thumbnail
		if (Thumbnail.IsValid())
			Render2D::DrawSprite(Thumbnail, rectangle);
		else
			Render2D::FillRectangle(rectangle, Colors::Black);
	}

	void ContentItem::DrawThumbnail(Rectangle rectangle, bool shadow)
	{
		// Draw shadow
		if (shadow)
		{
			const float thumbnailInShadowSize = 50.0f;
			Rectangle shadowRect = rectangle.MakeExpanded((DefaultThumbnailSize - thumbnailInShadowSize) * rectangle.GetWidth() / DefaultThumbnailSize * 1.3f);
			if (!_shadowIcon.IsValid())
				_shadowIcon = EditorIcons::AssetShadow128;
			Render2D::DrawSprite(_shadowIcon, shadowRect);
		}

		// Draw thumbnail
		if (Thumbnail.IsValid())
			Render2D::DrawSprite(Thumbnail, rectangle);
		else
			Render2D::FillRectangle(rectangle, Colors::Black);
	}

	void ContentItem::AddReference(IContentItemOwner* obj)
	{
		ENGINE_ASSERT(obj != nullptr);
		ENGINE_ASSERT(!_references.Contains(obj))
		_references.Add(obj);

		// Check if need to generate preview
		if (_references.Count() == 1 && !Thumbnail.IsValid())
		{
			RequestThumbnail();
		}
	}

	void ContentItem::RemoveReference(IContentItemOwner* obj)
	{
		if (_references.Remove(obj))
		{
			// Check if need to release the preview
			if (_references.Count() == 0 && Thumbnail.IsValid())
			{
				ReleaseThumbnail();
			}
		}
	}

	void ContentItem::OnDelete()
	{
		// Fire event
		while (_references.Count() > 0)
		{
			IContentItemOwner* reference = _references[0];
			reference->OnItemDeleted(this);
			RemoveReference(reference);
		}

		// Release thumbnail
		if (Thumbnail.IsValid())
		{
			ReleaseThumbnail();
		}
	}

	void ContentItem::NavigationFocus()
	{
		Control::NavigationFocus();

		if (IsFocused)
		{
			ContentView* view = TypeTryCast<ContentView>(Parent);
			if (view != nullptr)
			{
				view->Select(this);
			}
		}
	}

	void ContentItem::Draw()
	{
		Float2 size = Size;
		Style* style = Style::Current;
		ContentView* view = TypeTryCast<ContentView>(Parent);
		bool isSelected = view->IsSelected(this);
		Rectangle clientRect = Rectangle(Float2::Zero, size);
		Rectangle textRect = TextRectangle;
		Rectangle thumbnailRect;
		TextAlignment nameAlignment;
		switch (view->ViewType)
		{
			case ContentViewType::Tiles:
			{
				float thumbnailSize = size.x;
				thumbnailRect = Rectangle(0, 0, thumbnailSize, thumbnailSize);
				nameAlignment = TextAlignment::Center;

				if (TypeIs<ContentFolder>(this))
				{
					// Small shadow
					Rectangle shadowRect = Rectangle(2, 2, clientRect.GetWidth() + 1, clientRect.GetHeight() + 1);
					Color color = Colors::Black.AlphaMultiplied(0.2f);
					Render2D::FillRectangle(shadowRect, color);
					Render2D::FillRectangle(clientRect, style->Background.RGBMultiplied(1.25f));

					float accentHeight = 2 * view->ViewScale;
					Rectangle barRect = Rectangle(0, thumbnailRect.GetHeight() - accentHeight, clientRect.GetWidth(), accentHeight);
					Render2D::FillRectangle(barRect, Colors::DimGray);

					if (isSelected)
						Render2D::FillRectangle(clientRect, Parent->ContainsFocus() ? style->BackgroundSelected : style->LightBackground);
					else if (IsMouseOver)
						Render2D::FillRectangle(clientRect, style->BackgroundHighlighted);

					DrawThumbnail(thumbnailRect, false);
				}
				else
				{
					// Small shadow
					Rectangle shadowRect = Rectangle(2, 2, clientRect.GetWidth() + 1, clientRect.GetHeight() + 1);
					Color color = Colors::Black.AlphaMultiplied(0.2f);
					Render2D::FillRectangle(shadowRect, color);

					Render2D::FillRectangle(clientRect, style->Background.RGBMultiplied(1.25f));
					Render2D::FillRectangle(TextRectangle, style->LightBackground);

					DrawThumbnail(thumbnailRect, false);

					float accentHeight = 2 * view->ViewScale;
					Rectangle barRect = Rectangle(0, thumbnailRect.GetHeight() - accentHeight, clientRect.GetWidth(), accentHeight);
					Render2D::FillRectangle(barRect, Colors::DimGray);

					if (isSelected)
					{
						Render2D::FillRectangle(textRect, Parent->ContainsFocus() ? style->BackgroundSelected : style->LightBackground);
						Render2D::DrawRectangle(clientRect, Parent->ContainsFocus() ? style->BackgroundSelected : style->LightBackground);
					}
					else if (IsMouseOver)
					{
						Render2D::FillRectangle(textRect, style->BackgroundHighlighted);
						Render2D::DrawRectangle(clientRect, style->BackgroundHighlighted);
					}
				}
				break;
			}
			case ContentViewType::List:
			{
				float thumbnailSize = size.y - 2 * DefaultMarginSize;
				thumbnailRect = Rectangle(DefaultMarginSize, DefaultMarginSize, thumbnailSize, thumbnailSize);
				nameAlignment = TextAlignment::Near;

				if (isSelected)
					Render2D::FillRectangle(clientRect, Parent->ContainsFocus() ? style->BackgroundSelected : style->LightBackground);
				else if (IsMouseOver)
					Render2D::FillRectangle(clientRect, style->BackgroundHighlighted);

				DrawThumbnail(thumbnailRect);
				break;
			}
		}

		// Draw short name
		Render2D::PushClip(textRect);
		Render2D::RenderText(style->FontMedium, ShowFileExtension || view->ShowFileExtensions ? FileName : ShortName, textRect, style->TextColor,
			nameAlignment, TextAlignment::Center, TextWrapping::WrapWords, 1.0f, 0.95f);
		Render2D::PopClip();
	}


	bool ContentItem::OnMouseDown(Float2 location, MouseButton button)
	{
		Focus();

		if (button == MouseButton::Left)
		{
			// Cache data
			_isMouseDown = true;
			_mouseDownStartPos = location;
		}

		return true;
	}

	bool ContentItem::OnMouseUp(Float2 location, MouseButton button)
	{
		if (button == MouseButton::Left && _isMouseDown)
		{
			// Clear flag
			_isMouseDown = false;

			// Fire event
			ContentView* otherItem = TypeTryCast<ContentView>(this);
			if (otherItem != nullptr)
			{
				otherItem->OnItemClick(this);
			}
		}

		return Control::OnMouseUp(location, button);
	}

	bool ContentItem::OnMouseDoubleClick(Float2 location, MouseButton button)
	{
		Focus();

		ContentView* otherItem = TypeTryCast<ContentView>(Parent);
		if (otherItem != nullptr)
		{
			// Open
			otherItem->OnItemDoubleClick(this);
		}

		return true;
	}

	void ContentItem::OnMouseMove(Float2 location)
	{
		// Check if start drag and drop
		if (_isMouseDown && Float2::Distance(_mouseDownStartPos, location) > 10.0f)
		{
			// Clear flag
			_isMouseDown = false;

			// Start drag drop
			DoDrag();
		}
	}

	void ContentItem::OnMouseLeave()
	{
		// Check if start drag and drop
		if (_isMouseDown)
		{
			// Clear flag
			_isMouseDown = false;

			// Start drag drop
			DoDrag();
		}

		Control::OnMouseLeave();
	}

	void ContentItem::OnSubmit()
	{
		// Open
		/*AssetView* otherItem = TypeTryCast<AssetView>(other);
		(Parent as ContentView).OnItemDoubleClick(this);*/

		Control::OnSubmit();
	}

	int ContentItem::Compare(const Control* other) const
	{
		const ContentItem* otherItem = TypeTryCast<ContentItem>(other);
		if (otherItem != nullptr)
		{
			if (otherItem->IsFolder)
				return 1;
			return StringUtils::Compare(ShortName.Get(), otherItem->ShortName.Get());
		}

		return Control::Compare(other);
	}

	void ContentItem::OnDestroy()
	{
		// Fire event
		while (_references.Count() > 0)
		{
			IContentItemOwner* reference = _references[0];
			reference->OnItemDispose(this);
			RemoveReference(reference);
		}

		// Release thumbnail
		if (Thumbnail.IsValid())
		{
			ReleaseThumbnail();
		}

		Control::OnDestroy();
	}


	ContentItem::ContentItem(StringView path) : Control(0, 0, DefaultWidth, DefaultHeight)
	{
		// Set path
		Path = path;
		FileName = FileSystem::GetFileName(path);
		ShortName = FileSystem::GetFileNameWithoutExtension(path);
	}

	void ContentItem::RequestThumbnail()
	{
		EditorApp::Ins().uiModule->thumbnails->RequestPreview(this);
	}

	void ContentItem::ReleaseThumbnail()
	{
		// Simply unlink sprite
		Thumbnail = SpriteHandle::Invalid;
	}

	void ContentItem::OnReimport()
	{
		for (int i = 0; i < _references.Count(); i++)
			_references[i]->OnItemReimported(this);
		RefreshThumbnail();
	}

	void ContentItem::DoDrag()
	{
		if (!CanDrag)
			return;

		DragData* data = nullptr;

		// Check if is selected
		ContentView* view;
		if (TypeTryCast<ContentView>(Parent, view) && view->IsSelected(this))
		{
			// Drag selected item
			data = DragItems::GetDragData(view->GetSelection());
		}
		else
		{
			// Drag single item
			data = DragItems::GetDragData(this);
		}

		// Start drag operation
		DoDragDrop(data);
	}

	void ContentItem::OnBuildTooltipText(StringBuilder sb)
	{
		/*sb.Append("Type: ").Append(TypeDescription).AppendLine();
		if (FileSystem::FileExists(Path))
		{
			sb.Append("Size: ").Append(Utilities.Utils.FormatBytesCount((int)new FileInfo(Path).Length)).AppendLine();
		}
		sb.Append("Path: ").Append(Utilities.Utils.GetAssetNamePathWithExt(Path)).AppendLine();*/
	}

	bool ContentItem::__GetCanRename()
	{
		return true;
	}

	bool ContentItem::__GetCanDrag()
	{
		return Root != nullptr;
	}

	bool ContentItem::__GetExists()
	{
		return FileSystem::FileExists(Path);
	}

	SpriteHandle ContentItem::__GetDefaultThumbnail()
	{
		return SpriteHandle::Invalid;
	}

	bool ContentItem::__GetHasDefaultThumbnail()
	{
		return false;
	}
} // SE

File: Src/Engine/Editor/Resource/Items/ContentItem.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using SE.GUI;

namespace SE.Editor
{
    public enum ContentItemType
    {
        Asset,
        Folder,
        Script,
        Scene,
        Other,
    }

    public enum ContentItemSearchFilter
    {
        Model,
        SkinnedModel,
        Material,
        Texture,
        Scene,
        Prefab,
        Script,
        Audio,
        Animation,
        Json,
        Particles,
        Shader,
        Other,
    }

    public enum ContentFolderType
    {
        Content,
        Source,
        Other,
    }

    public interface IContentItemOwner
    {
        void OnItemDeleted(ContentItem item);
        void OnItemRenamed(ContentItem item);
        void OnItemReimported(ContentItem item);
        void OnItemDispose(ContentItem item);
    }

    /// <summary>
    /// Managed presentation model for one entry in the editor content browser.
    /// It deliberately stores paths and managed state only; native assets are
    /// addressed through stable ids instead of raw C++ pointers.
    /// </summary>
    public abstract class ContentItem : Control
    {
        public const int DefaultMarginSize = 4;
        public const int DefaultTextHeight = 24;
        public const int DefaultThumbnailSize = 64;
        public const int DefaultWidth = DefaultThumbnailSize + 2 * DefaultMarginSize;
        public const int DefaultHeight = DefaultThumbnailSize + 2 * DefaultMarginSize + DefaultTextHeight;

        private readonly HashSet<IContentItemOwner> m_Owners = new();
        private bool m_MouseDown;
        private Float2 m_MouseDownStart;

        protected ContentItem(string path)
            : base(new Rectangle(0, 0, DefaultWidth, DefaultHeight))
        {
            UpdatePath(path);
        }

        public abstract ContentItemType ItemType { get; }
        public virtual ContentItemSearchFilter SearchFilter => ContentItemSearchFilter.Other;
        public bool IsAsset => ItemType == ContentItemType.Asset;
        public bool IsFolder => ItemType == ContentItemType.Folder;
        public virtual bool CanHaveChildren => IsFolder;
        public virtual bool CanRename => ParentFolder != null;
        public virtual bool CanDrag => ParentFolder != null;
        public virtual bool Exists => File.Exists(Path) || Directory.Exists(Path);
        public ContentFolder? ParentFolder { get; private set; }
        public string Path { get; private set; } = string.Empty;
        public string FileName { get; private set; } = string.Empty;
        public string ShortName { get; private set; } = string.Empty;
        public string NamePath { get; private set; } = string.Empty;
        public string Tooltip { get; protected set; } = string.Empty;
        public object? Thumbnail { get; internal set; }
        public bool ShowFileExtension { get; set; }
        public int ReferencesCount => m_Owners.Count;
        public virtual object? DefaultThumbnail => null;
        public bool HasDefaultThumbnail => DefaultThumbnail != null;
        public virtual string DisplayName => ShowFileExtension ? FileName : ShortName;

        public event Action<ContentItem>? DoubleClicked;
        public event Action<ContentItem>? Reloaded;

        public virtual void UpdatePath(string path)
        {
            ArgumentException.ThrowIfNullOrWhiteSpace(path);
            Path = path;
            FileName = System.IO.Path.GetFileName(path);
            ShortName = System.IO.Path.GetFileNameWithoutExtension(path);
            NamePath = ShortName;
            Name = DisplayName;
            UpdateTooltipText();
            OnPathChanged();
            Notify(static (owner, item) => owner.OnItemRenamed(item));
        }

        public virtual void RefreshThumbnail()
        {
            ThumbnailService.Instance.RequestPreview(this);
        }

        public virtual void UpdateTooltipText()
        {
            Tooltip = Path;
            TooltipText = Tooltip;
        }

        public virtual ContentItem? Find(string path)
        {
            return string.Equals(Path, path, StringComparison.OrdinalIgnoreCase) ? this : null;
        }

        public virtual bool Find(ContentItem item) => ReferenceEquals(this, item);

        public void AddReference(IContentItemOwner owner)
        {
            ArgumentNullException.ThrowIfNull(owner);
            m_Owners.Add(owner);
        }

        public void RemoveReference(IContentItemOwner owner)
        {
            ArgumentNullException.ThrowIfNull(owner);
            m_Owners.Remove(owner);
        }

        public virtual void OnDelete()
        {
            Notify(static (owner, item) => owner.OnItemDeleted(item));
            ParentFolder?.RemoveChild(this);
        }

        public override bool OnMouseDown(Float2 location, MouseButton button)
        {
            if (button != MouseButton.Left)
                return base.OnMouseDown(location, button);

            m_MouseDown = true;
            m_MouseDownStart = location;
            FindContentView()?.OnItemClick(this);
            return true;
        }

        public override bool OnMouseUp(Float2 location, MouseButton button)
        {
            _ = location;
            if (button != MouseButton.Left || !m_MouseDown)
                return base.OnMouseUp(location, button);

            m_MouseDown = false;
            return true;
        }

        public override bool OnMouseDoubleClick(Float2 location, MouseButton button)
        {
            _ = location;
            if (button != MouseButton.Left)
                return base.OnMouseDoubleClick(location, button);

            FindContentView()?.OnItemDoubleClick(this);
            DoubleClicked?.Invoke(this);
            return true;
        }

        public override void OnMouseMove(Float2 location)
        {
            if (m_MouseDown && (location - m_MouseDownStart).LengthSquared > 16.0f)
                m_MouseDown = false;
            base.OnMouseMove(location);
        }

        protected virtual void OnPathChanged()
        {
        }

        protected virtual void OnReimport()
        {
            Reloaded?.Invoke(this);
            Notify(static (owner, item) => owner.OnItemReimported(item));
        }

        protected override void OnDispose()
        {
            ParentFolder?.RemoveChild(this);
            Notify(static (owner, item) => owner.OnItemDispose(item));
            m_Owners.Clear();
            base.OnDispose();
        }

        internal void SetParentFolder(ContentFolder? value)
        {
            ParentFolder = value;
            OnParentFolderChanged();
        }

        protected virtual void OnParentFolderChanged()
        {
        }

        protected void NotifyReimported() => OnReimport();

        private void Notify(Action<IContentItemOwner, ContentItem> callback)
        {
            var owners = m_Owners.ToArray();
            foreach (IContentItemOwner owner in owners)
            {
                callback(owner, this);
            }
        }

        private ContentView? FindContentView()
        {
            for (Control? control = Parent; control != null; control = control.Parent)
            {
                if (control is ContentView view)
                    return view;
            }

            return null;
        }
    }

    public class ContentFolder : ContentItem
    {
        private readonly List<ContentItem> m_Children = new();

        public ContentFolder(ContentFolderType folderType, string path)
            : base(path)
        {
            FolderType = folderType;
        }

        public override ContentItemType ItemType => ContentItemType.Folder;
        public ContentFolderType FolderType { get; }
        public bool CanHaveAssets => FolderType == ContentFolderType.Content;
        public ContentTreeNode? Node { get; internal set; }
        public IReadOnlyList<ContentItem> Children => m_Children;
        public override bool CanRename => ParentFolder != null;
        public override bool CanDrag => ParentFolder != null;

        public ContentItem AddChild(ContentItem item)
        {
            ArgumentNullException.ThrowIfNull(item);
            if (ReferenceEquals(item, this))
                throw new InvalidOperationException("A folder cannot contain itself.");
            if (item.ParentFolder == this)
                return item;

            item.ParentFolder?.RemoveChild(item);
            m_Children.Add(item);
            item.SetParentFolder(this);
            return item;
        }

        public bool RemoveChild(ContentItem item)
        {
            if (!m_Children.Remove(item))
                return false;

            item.SetParentFolder(null);
            return true;
        }

        public ContentItem? FindChild(string path)
        {
            return m_Children.Find(item => string.Equals(item.Path, path, StringComparison.OrdinalIgnoreCase));
        }

        public bool ContainsChild(string path) => FindChild(path) != null;

        public override void UpdatePath(string path)
        {
            string previousPath = Path;
            base.UpdatePath(path);
            if (string.IsNullOrEmpty(previousPath))
                return;

            foreach (ContentItem child in m_Children)
            {
                string relative = System.IO.Path.GetRelativePath(previousPath, child.Path);
                child.UpdatePath(System.IO.Path.Combine(path, relative));
            }
        }

        public override ContentItem? Find(string path)
        {
            ContentItem? direct = base.Find(path);
            if (direct != null)
                return direct;

            foreach (ContentItem child in m_Children)
            {
                ContentItem? result = child.Find(path);
                if (result != null)
                    return result;
            }

            return null;
        }

        public override bool Find(ContentItem item)
        {
            if (base.Find(item))
                return true;

            foreach (ContentItem child in m_Children)
            {
                if (child.Find(item))
                    return true;
            }

            return false;
        }

        protected override void OnDispose()
        {
            foreach (ContentItem child in m_Children.ToArray())
            {
                child.Dispose();
            }
            m_Children.Clear();
            base.OnDispose();
        }
    }

    public class AssetItem : ContentItem
    {
        public AssetItem(string path, string typeId, Guid id)
            : base(path)
        {
            TypeId = typeId ?? string.Empty;
            Id = id;
        }

        public override ContentItemType ItemType => ContentItemType.Asset;
        public Guid Id { get; }
        public string TypeId { get; }
        public bool IsLoaded { get; private set; }
        public override ContentItemSearchFilter SearchFilter => ContentItemSearchFilter.Other;

        public virtual void SetLoaded(bool value)
        {
            IsLoaded = value;
        }

        public virtual void Reload()
        {
            NotifyReimported();
        }
    }

    public class BinaryAssetItem : AssetItem
    {
        public BinaryAssetItem(string path, string typeId, Guid id, string? sourcePath = null)
            : base(path, typeId, id)
        {
            SourcePath = sourcePath;
        }

        public string? SourcePath { get; private set; }

        public bool TryGetImportPath(out string? path)
        {
            path = SourcePath;
            return !string.IsNullOrEmpty(path);
        }

        public void SetImportPath(string? path)
        {
            SourcePath = path;
        }
    }

    public sealed class FileItem : ContentItem
    {
        public FileItem(string path)
            : base(path)
        {
        }

        public override ContentItemType ItemType => ContentItemType.Other;
        public override ContentItemSearchFilter SearchFilter => ContentItemSearchFilter.Other;
    }

    public sealed class JsonAssetItem : AssetItem
    {
        public JsonAssetItem(string path, string typeId, Guid id)
            : base(path, typeId, id)
        {
        }

        public override ContentItemSearchFilter SearchFilter => ContentItemSearchFilter.Json;
    }

    public sealed class SceneItem : AssetItem
    {
        public SceneItem(string path, Guid id)
            : base(path, "SceneAsset", id)
        {
        }

        public override ContentItemType ItemType => ContentItemType.Scene;
        public override ContentItemSearchFilter SearchFilter => ContentItemSearchFilter.Scene;
    }

    public sealed class NewItem : ContentItem
    {
        public NewItem(string path)
            : base(path)
        {
        }

        public override ContentItemType ItemType => ContentItemType.Other;
        public override bool Exists => false;
        public override bool CanDrag => false;
    }
}
